wallet/db.c

```diff
@@ -331,7 +331,72 @@ char *dbmigrations[] = {
     NULL,
 };
 
-sqlite3_stmt *db_prepare_(const char *caller, struct db *db, const char *query)
+/* Leak tracking. */
+#if DEVELOPER
+/* We need a global here, since caller has no context.  Yuck! */
+static struct list_head db_statements = LIST_HEAD_INIT(db_statements);
+
+struct db_statement {
+	struct list_node list;
+	sqlite3_stmt *stmt;
+	const char *origin;
+};
+
+static struct db_statement *find_statement(sqlite3_stmt *stmt)
+{
+	struct db_statement *i;
+
+	list_for_each(&db_statements, i, list) {
+		if (i->stmt == stmt)
+			return i;
+	}
+	return NULL;
+}
+
+void db_assert_no_outstanding_statements(void)
+{
+	struct db_statement *dbstat;
+
+	dbstat = list_top(&db_statements, struct db_statement, list);
+	if (dbstat)
+		fatal("Unfinalized statement %s", dbstat->origin);
+}
+
+static void dev_statement_start(sqlite3_stmt *stmt, const char *origin)
+{
+	struct db_statement *dbstat = tal(NULL, struct db_statement);
+	dbstat->stmt = stmt;
+	dbstat->origin = origin;
+	list_add(&db_statements, &dbstat->list);
+}
+
+static void dev_statement_end(sqlite3_stmt *stmt)
+{
+	struct db_statement *dbstat = find_statement(stmt);
+	list_del_from(&db_statements, &dbstat->list);
+	tal_free(dbstat);
+}
+#else
+static void dev_statement_start(sqlite3_stmt *stmt, const char *origin)
```

Comment:
Any particular reason we leave these stubs around? There are very few places we actually call them, so we could just add a `DEVELOPER=1` check in those places.

Reply:
Then again I guess they'll be optimized out by the compiler.

Reply:
Yes, it reduces the number of `#ifdef`..`#endif` blocks in the code, and as you indeed mention, they will tend to get optimized out when empty, especially as they are `static` also.